### PR TITLE
Filipchik_Kirill

### PR DESCRIPTION
**Pull request by filipchikkirill**

### CODE REVIEW

**Comment by NikolaevArtem** [comment]
Please merge with master

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,37 @@
 # Project exclude paths
 /.gradle/
 /build/
-/build/classes/java/main/
\ No newline at end of file
+/build/classes/java/main/
```
Comment (NikolaevArtem) [inline .gitignore]: Please merge with master

File: src/main/java/homework_1/Main.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package homework_1;
+
+public class Main {
+    public static void main(String... args) {
+        if (args.length == 0) return;
+        StringBuilder strBuilder = new StringBuilder();
+        for (String arg :
+                args) {
+            if (arg.equalsIgnoreCase("ошибка")) {
+                System.err.println("Тревога!");
```
Comment (NikolaevArtem) [inline src/main/java/homework_1/Main.java]: Doesn't work as expected

File: src/main/java/homework_1/Main.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package homework_1;
+
+public class Main {
+    public static void main(String... args) {
+        if (args.length == 0) return;
```
Comment (NikolaevArtem) [inline src/main/java/homework_1/Main.java]: Not needed

File: src/main/java/homework_1/Main.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        StringBuilder strBuilder = new StringBuilder();
+        for (String arg :
+                args) {
+            if (arg.equalsIgnoreCase("ошибка")) {
+                System.err.println("Тревога!");
+                return;
+            } else {
+                strBuilder
```
Comment (NikolaevArtem) [inline src/main/java/homework_1/Main.java]: Looks complicated, better simplify or move to a method

File: src/main/java/homework_1/Main.java [hunk 5]
```diff
@@ -1,9 +1,39 @@
 package homework_1;
 
 public class Main {
+  public static final String redColorOutString = "\u001B[31m";
+  public static final String closeColorOutString = "\u001B[0m";
+    public static void main(String... args) {
+        for (String arg :
+                args) {
+            if (arg. equals("ошибка")) {
```
Comment (NikolaevArtem) [inline src/main/java/homework_1/Main.java]: Opt: please fix style according code conventions

File: src/main/java/homework_2/pyramid_printer/Main.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package homework_2.pyramid_printer;
+
+import homework_2.utils.Executable;
+
+public class Main {
+
+    public static void main(String[] args) {
+        Executable pyramidPrinter = new PyramidPrinter();
+        pyramidPrinter.run();
```
Comment (NikolaevArtem) [inline src/main/java/homework_2/pyramid_printer/Main.java]: Opt: could be simplified to 
```suggestion
        new PyramidPrinter().run();
```

File: src/main/java/homework_2/random_chars_table/RandomCharsTable.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package homework_2.random_chars_table;
+
+import homework_2.utils.App;
+
+import java.util.Random;
+
+public class RandomCharsTable extends App {
```
Comment (NikolaevArtem) [inline src/main/java/homework_2/random_chars_table/RandomCharsTable.java:7]: Somehow, without letters it removes the line
image

Comment (NikolaevArtem) [reply src/main/java/homework_2/random_chars_table/RandomCharsTable.java:7]: Good job! All tests are passed, works as expected. For style, it's good that you tried most of the things we learn, but in prod, I don't recommend doing it, as simplicity is usually the main goal

Comment (NikolaevArtem) [reply src/main/java/homework_2/random_chars_table/RandomCharsTable.java:7]: Opt: make homework simpler to undertand where possible

File: src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        printOut();
+        removeFromOutput("Введите данные:");
+        for (int i = 0; i < ROWS; i++)
+        assertTrue(getOutputLines()[i].matches("(\\|\\s[A-Z]\\s){" + COLUMNS + "}\\|"));
+    }
+
+    // дальше тест уродлив, но как иначе распарсить эти строки и сравнить все я не знаю
+    // за то покрывает все варианты.
+    @RepeatedTest(10)
+    void InputRandomCheckStrategy() {
```
Comment (NikolaevArtem) [inline src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java:90]: It seems it does the trick, but possible bugs could be present. If possible, simplify :)

File: src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    @Test
+    void InputEmptyArgument() {
+        setInput("");
+        assertThrows(NoSuchElementException.class, () -> new RandomCharsTable().run());
+    }
+
+    /*
+        for this test use new RandomCharsTable().setDebug(true).run();
```
Comment (NikolaevArtem) [inline src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java:30]: If think, once you'll understand how to check random letters, you'll not need debug mode. Good idea anyway

File: src/main/java/homework_3/ImmutableClass.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+Data members in the class must be declared as private
+Data members in the class must be declared as final
+A parameterized constructor should initialize all the fields performing a deep copy
+Deep Copy of objects should be performed in the getter methods
+No setters
+ */
+public final class ImmutableClass {
+    private final String aString;
+    private final int anInt;
+    private final List<String> aList;
```
Comment (NikolaevArtem) [inline src/main/java/homework_3/ImmutableClass.java:17]: Opt: Variables usually are not named this way

File: src/main/java/homework_3/ImmutableClass.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+/*
+The class must be declared as final
+Data members in the class must be declared as private
+Data members in the class must be declared as final
+A parameterized constructor should initialize all the fields performing a deep copy
+Deep Copy of objects should be performed in the getter methods
+No setters
+ */
+public final class ImmutableClass {
```
Comment (NikolaevArtem) [inline src/main/java/homework_3/ImmutableClass.java:14]: Good

File: src/main/java/homework_2/random_chars_table/RandomCharsTable.java (hunk 7 above)
Comment (filipchikkirill) [reply src/main/java/homework_2/random_chars_table/RandomCharsTable.java:7]: If there are no letters suitable for the strategy, then the string - "* Strategy * letters" - should not be displayed? Did I understand correctly?
Of course, I added inheritance for the sake of inheritance. In real tasks, this is not necessary, I understand that.

File: src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import static homework_4.custom_file_reader.CustomFileReader.*;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+class CustomFileReaderTest extends UnitBase {
+
+    @Test
+    void TestRun1() throws IOException {
+        run1();
+        assertFalse(getOutput().contains("."));
+        assertFalse(getOutput().contains(","));
```
Comment (tmandreeva1) [inline src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:18]: Probably, what should you really test, is edge cases: what happens if file is empty? Or doesn't contain any dots or commas? Also, since you have setter for a file path in your CustomFileReader - what if we set a path of a non-existent file? 
And all of this test cases can be applied to all of the reading methods.

Comment (filipchikkirill) [reply src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:18]: "реализовать приложение, которое считывает данные из файла, и печатает в консоль, но уже без запятых и точек."
If there are no commas or dots in the file, or the file is empty, then these tests will pass perfectly.
The output should not contain dots and commas - this is the condition of the task. This is verified by tests.

"Файл - тот, что вы даете. Другого файла не ожидается. Все проще) "
Thanks for the comment, removed the setter.

Comment (filipchikkirill) [reply src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:18]: Add test - "removed only commas and dots" is a good idea

Comment (tmandreeva1) [reply src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:18]: _> If there are no commas or dots in the file, or the file is empty, then these tests will pass perfectly._ 

But what if you change your run method implementation in future? Test may fail in one situation and pass in others and that is not what is expected from tests. I mean:
1. The purpose of tests is to be able to add new code, or remove, or refactor, or whatever with being assure that you're not impacting the code you weren't plan to impact, that other app features work correctly. 
2. Tests need to be atomic, which means that one test check one test case and if it fails, it's crystal clear what is wrong. If your current tests fail, you certainly will understand which method is not working as expected (run1 or run2 etc), but you will need to test your app manually to get the concrete cause of test fail. Which will cost you some time. (test passed with ok and ordinary and expected file and test passed with empty file - these are two different test cases, because they have different data given).

I know, that hw4 wasn't about testing, nor about changing app in future (which, by the way, you never know), but your answer made me feel, that you're not properly understand testing, which might be important sometime in your career. You could search for some reliable sources about testing, if you want, it's quite good to know.
Also, whatever the task is, testing edge cases is must-have, in my opinion. 

Whatever, it's just the thoughts:)

Comment (tmandreeva1) [reply src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:18]: > "Файл - тот, что вы даете. Другого файла не ожидается. Все проще) " 

File - file name - could be the same, contents might differ

File: src/main/java/homework_4/custom_file_reader/Main.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package homework_4.custom_file_reader;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+public class Main {
+    public static void main(String[] args) {
+        try {
```
Comment (ArtemNikolaev1) [inline src/main/java/homework_4/custom_file_reader/Main.java]: 1. Better have one try-with-resources per one stream. So it's better to have 4 try/catch in run() methods.
2. You don't close streams

Comment (filipchikkirill) [reply src/main/java/homework_4/custom_file_reader/Main.java]: Done.

File: src/main/java/homework_3/ImmutableClass.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    private final String aString;
+    private final int anInt;
+    private final List<String> aList;
+
+    public ImmutableClass(List<String> aList) {
+        this.aString = "default";
+        this.anInt = 0;
+        this.aList = aList;
```
Comment (NikolaevArtem) [inline src/main/java/homework_3/ImmutableClass.java]: Actually, I see now your class is mutable. A parameterized constructor should initialize all the fields performing a deep copy

File: src/main/java/homework_3/ImmutableClass.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    public int getAnInt() {
+        return anInt;
+    }
+
+    public List<String> getaList() {
+        return new ArrayList<>(aList);
+    }
+
+    public ImmutableClass getNew(List<String> aList) {
+        return new ImmutableClass(aString, anInt, aList);
```
Comment (NikolaevArtem) [inline src/main/java/homework_3/ImmutableClass.java]: Same here

File: src/main/java/homework_4/custom_file_reader/Main.java [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package homework_4.custom_file_reader;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+public class Main {
+    public static void main(String[] args) {
+        CustomFileReader.run();
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/custom_file_reader/Main.java:8]: Looks strange 
image

File: src/main/java/homework_4/custom_file_reader/CustomFileReader.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    public static void run1() throws IOException {
+        Scanner scanner = new Scanner(getFileURL().openStream()).useDelimiter("\n");;
+        scanner.forEachRemaining(FUNCTION);
+        scanner.close();
+    }
+
+    public static void run2() throws IOException {
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(getFileURL().openStream()));
+        bufferedReader.lines().forEach(FUNCTION);
+        bufferedReader.close();
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/custom_file_reader/CustomFileReader.java]: Now you close streams, but not where you should. Better use try-with-resources, you'll avoid this mistake

Comment (NikolaevArtem) [reply src/main/java/homework_4/custom_file_reader/CustomFileReader.java]: In case you get an exception on 48th line, bufferedReader will not be closed

File: src/main/java/homework_4/singleton/Singleton.java [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package homework_4.singleton;
+
+public final class Singleton {
+    private Singleton() {}
+
+    private static class Holder {
+        static final Singleton INSTANCE = new Singleton();
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/singleton/Singleton.java:7]: Opt: why not just have static variable?

File: src/main/java/homework_4/сustom_annotation/SimpleClassWithNameField.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    private String name;
+
+    @CustomAnnotation("test name string")
+    public SimpleClassWithNameField(String name) {
+        Constructor<? extends SimpleClassWithNameField> constructor;
+        try {
+            constructor = this.getClass().getConstructor(String.class);
+            CustomAnnotation annotation = constructor.getAnnotation(CustomAnnotation.class);
+            if (annotation != null) this.name = annotation.value();
+            if (!name.isEmpty()) this.name = name;
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/сustom_annotation/SimpleClassWithNameField.java]: Opt: 
1) let's not use if without brackets, this approach is prone to errors
2) I think, the idea you try to implement would look better if 'if-else' or ternary operator will be used

File: src/main/java/homework_4/сustom_annotation/CustomAnnotation.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.CONSTRUCTOR;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target(CONSTRUCTOR)
+public @interface CustomAnnotation {
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/сustom_annotation/CustomAnnotation.java:11]: Good!

File: src/main/java/homework_5/AppHW_5.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
+package homework_5;
+
+import homework_5.utils.MyReader;
+import homework_5.utils.StringChecker;
+import homework_5.utils.impl.MyReaderImpl;
+import homework_5.utils.impl.StringCheckerImpl;
+
+import java.util.NoSuchElementException;
+
+public class AppHW_5 {
```
Comment (NikolaevArtem) [inline src/main/java/homework_5/AppHW_5.java:10]: Opt: in general idea to separate user input class is a good one, but we should not forget that abstractions are made when they simplify reading and fixing the code, not make it less readable

File: src/main/java/homework_5/custom_regex_matcher/CustomRegexMatcher.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    public CustomRegexMatcher() {
+        super(PATTERN_STRING);
+    }
+
+    public String run() {
+        readData();
+        String result = String.valueOf(check());
+        System.out.println(result);
+        closeReader();
+        return result;
```
Comment (NikolaevArtem) [inline src/main/java/homework_5/custom_regex_matcher/CustomRegexMatcher.java:17]: Opt: as example here it's pretty unclear what is happening. Void methods usually are not recommended too

File: src/main/java/homework_6/map_problems_generator/problem_impl/GettingProblem.java [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+public class GettingProblem <T> extends NoProblem <T> {
+
+    public GettingProblem(T value) {
+        super(value);
+    }
+
+    @Override
+    public int hashCode() {
+        return new Random().nextInt();
```
Comment (NikolaevArtem) [inline src/main/java/homework_6/map_problems_generator/problem_impl/GettingProblem.java:13]: I reread the task description, and I agree it's unclear there. But the task I had in mind is when you put mutable key, change it's field, and try to get value -> you can't, because hash is now different, even though hash was written according contracts.

File: src/main/java/homework_6/map_problems_generator/problem_impl/Wrapper.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
+package homework_6.map_problems_generator.problem_impl;
+
+import homework_6.map_problems_generator.Problem;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Wrapper <T> implements Problem<T> {
+    private final Problem<T> innerProblem;
+    private final List<String> event = new ArrayList<>();
```
Comment (NikolaevArtem) [inline src/main/java/homework_6/map_problems_generator/problem_impl/Wrapper.java:10]: Good implementation for generating events!

File: src/main/java/homework_7/Main.java [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+import homework_7.kitten_to_cat_function.Cat;
+import homework_7.kitten_to_cat_function.Kitten;
+import homework_7.kitten_to_cat_function.KittenToCatFunction;
+
+public class Main {
+
+    public static void main(String[] args) {
+        Kitten kitten = new Kitten("Murzik", 1);
+        KittenToCatFunction function = (k) -> new Cat(k.getName(), k.getAge() + 4);
```
Comment (NikolaevArtem) [inline src/main/java/homework_7/Main.java:11]: Opt: you could add some more complicated logic to lambda, as example - converting one type to another (like, favourite toy to favourite color or smth like this)

File: src/main/java/course_project/sea_battle/SeaBattleApp.java [hunk 26]
```diff
@@ -0,0 +1,9 @@
+package course_project.sea_battle;
+
+import course_project.sea_battle.controller.impl.GameControllerImpl;
+
+public class SeaBattleApp {
```
Comment (NikolaevArtem) [inline src/main/java/course_project/sea_battle/SeaBattleApp.java:6]: Opt: The field is quite narrow, i think it would look better with more space between cells by width
image

Comment (NikolaevArtem) [reply src/main/java/course_project/sea_battle/SeaBattleApp.java:6]: The game seems not working. I see you implemented PvP, but I play only for 1 Player. I don't see field for 2nd, don't see his name anywhere. Don't know how to shoot for 2nd. 
And please introduce automatic ship placement, testing will be much easier

Comment (NikolaevArtem) [reply src/main/java/course_project/sea_battle/SeaBattleApp.java:6]: Good! Nice architecture, SOLID is followed. Code is pretty readable. Good Java Core and Java libraries usage. 

To make better: interface could be refactored. Also check correct working of the game.

Comment (NikolaevArtem) [reply src/main/java/course_project/sea_battle/SeaBattleApp.java:6]: Also nice approach to testing, test cases are readable

File: src/test/java/homework_4/singleton/SingletonTest.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+class SingletonTest {
+
+    @Test
+    void getInstance() {
+        Singleton singleton1 = Singleton.getInstance();
+        Singleton singleton2 = Singleton.getInstance();
+
+        assertEquals(singleton1, singleton2);
```
Comment (NikolaevArtem) [inline src/test/java/homework_4/singleton/SingletonTest.java:14]: Opt: they can be equal, but be different objects, right?

**Review by NikolaevArtem — CHANGES_REQUESTED** [review]
Nice! The code needs small fixes, but except that - good job! The architecture is nice, and I see you exercised with inheritance/generics, which is cool. But remember that it's nice for learning purposes, but production code, at best, should be readable and simple first. Usually, when there's a choice between SOLID or readability/simplicity, I choose second. IMHO

File: src/main/java/homework_4/singleton/Singleton.java (hunk 18 above)
Comment (filipchikkirill) [reply src/main/java/homework_4/singleton/Singleton.java:7]: Lazy init.

File: src/main/java/homework_4/custom_file_reader/Main.java (hunk 16 above)
Comment (filipchikkirill) [reply src/main/java/homework_4/custom_file_reader/Main.java:8]: How do I get this error? 
remind, i don't use windows

File: src/main/java/homework_3/ImmutableClass.java (hunk 14 above)
Comment (filipchikkirill) [reply src/main/java/homework_3/ImmutableClass.java]: done here

File: src/main/java/homework_3/ImmutableClass.java (hunk 15 above)
Comment (filipchikkirill) [reply src/main/java/homework_3/ImmutableClass.java]: I don't understand what the problem is, I redid it just in case...

File: src/main/java/homework_4/custom_file_reader/CustomFileReader.java (hunk 17 above)
Comment (filipchikkirill) [reply src/main/java/homework_4/custom_file_reader/CustomFileReader.java]: done. Add try with res

File: src/main/java/homework_4/singleton/Singleton.java (hunk 18 above)
Comment (filipchikkirill) [reply src/main/java/homework_4/singleton/Singleton.java:7]: The JVM defers initializing the InstanceHolder class until it is actually used, and because the Singleton is initialized with a static initializer, no additional synchronization is needed. The first call to getInstance by any thread causes Holder to be loaded and initialized, at which time the initialization of the Singleton happens through the static initializer.

File: src/main/java/homework_4/сustom_annotation/SimpleClassWithNameField.java (hunk 19 above)
Comment (filipchikkirill) [reply src/main/java/homework_4/сustom_annotation/SimpleClassWithNameField.java]: add brackets

File: src/main/java/homework_6/map_problems_generator/problem_impl/GettingProblem.java (hunk 23 above)
Comment (filipchikkirill) [reply src/main/java/homework_6/map_problems_generator/problem_impl/GettingProblem.java:13]: add this case(new class add)

File: src/main/java/homework_4/singleton/Singleton.java (hunk 18 above)
Comment (NikolaevArtem) [reply src/main/java/homework_4/singleton/Singleton.java:7]: Agree, done

File: src/test/java/homework_4/singleton/SingletonTest.java (hunk 27 above)
Comment (filipchikkirill) [reply src/test/java/homework_4/singleton/SingletonTest.java:14]: they share a common pointer. (added this check (IDE don't like this))

File: src/main/java/course_project/sea_battle/SeaBattleApp.java (hunk 26 above)
Comment (NikolaevArtem) [reply src/main/java/course_project/sea_battle/SeaBattleApp.java:6]: Good, automatic ship placement! I found some minor issues with the interface: 
1) When miss, there's no message to press enter to finish the turn. Actually, I didn't see this message anywhere... Confusing
2) In case i hit a ship, i need to press enter 2 times to be able to shoot again
3) No error message in case I already shot this cell
image
4) No message in case of destroyed ship 
image
5) And please make field representation wider :) 

The app now works correctly, I didn't find any bugs. And from the code perspective, the app looks great! 

To make better: improve interface:) 
The homework is approved

File: src/main/java/homework_4/custom_file_reader/CustomFileReader.java [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+        try {
+            run3();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        try {
+            run4();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (NikolaevArtem) [inline src/main/java/homework_4/custom_file_reader/CustomFileReader.java:37]: Opt: these try-catch now are either not needed or not in right place...

**Review by NikolaevArtem — APPROVED** [review]
Approved!
